### PR TITLE
Final bits of W3DView

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

**Comment by jonwil** [comment]
USA01 and GLA01 play through with no issues.

File: src/platform/w3dengine/client/w3dview.cpp [hunk 1]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+    Coord3D &pos = Get_Position();
+    float x = epicenter->x - pos.x;
+    float y = epicenter->y - pos.y;
+    float length = GameMath::Sqrt(x * x + y * y);
+
+    if (length <= g_theWriteableGlobalData->m_maxShakeRange) {
+        m_shakeIntensity += (1.0f - length / g_theWriteableGlobalData->m_maxShakeRange) * intensity;
+
+        if (m_shakeIntensity > g_theWriteableGlobalData->m_maxShakeIntensity) {
+            m_shakeIntensity = 3.0f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:2184]: 3.0? Not g_theWriteableGlobalData->m_maxShakeIntensity ?

File: src/platform/w3dengine/client/w3dview.cpp [hunk 2]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+            intensity = g_theWriteableGlobalData->m_shakeCineInsaneIntensity;
+            break;
+        default:
+            break;
+    }
+
+    Coord3D &pos = Get_Position();
+    float x = epicenter->x - pos.x;
+    float y = epicenter->y - pos.y;
+    float length = GameMath::Sqrt(x * x + y * y);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: range

File: src/platform/w3dengine/client/w3dview.cpp [hunk 3]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+    Zoom_Camera(zoom, frames, in, out);
+    Pitch_Camera(1.0f, frames, in, out);
+}
+
+bool W3DView::Is_Camera_Movement_Finished()
+{
+    if (m_viewFilter == FT_VIEW_MOTION_BLUR_FILTER
+        && (m_viewFilterMode == FM_VIEW_MB_IN_AND_OUT_ALPHA || m_viewFilterMode == FM_VIEW_MB_IN_AND_OUT_STATURATE
+            || m_viewFilterMode == FM_8 || m_viewFilterMode == FM_VIEW_MB_OUT_ALPHA
+            || m_viewFilterMode == FM_VIEW_MB_IN_STATURATE || m_viewFilterMode == FM_VIEW_MB_OUT_STATURATE)) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:2138]: `STATURATE` is correct spelling? Not SATURATE?

Multiples.

File: src/platform/w3dengine/client/w3dview.cpp [hunk 4]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+        m_pcInfo.end_time_multiplier = final_multiplier;
+    }
+
+    if (m_doingRotateCamera) {
+        m_rcInfo.end_time_multiplier = final_multiplier;
+    } else if (m_doingMoveCameraOnWaypointPath) {
+        float total_length = 0.0f;
+
+        for (int i = 0; i < m_mcwpInfo.num_waypoints; i++) {
+            total_length = total_length + m_mcwpInfo.way_seg_length[i];
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: or +=

File: src/platform/w3dengine/client/w3dview.cpp [hunk 5]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+            float length = v1.Length();
+
+            if (length >= 0.1f) {
+                float angle = GameMath::Acos(v1.X / length);
+
+                if (v1.Y < 0.0f) {
+                    angle = -angle;
+                }
+
+                angle -= 1.5707964f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Ideally all angle values are written with `DEG_TO_RADF()`

In this case `DEG_TO_RADF(90.f)`

File: src/platform/w3dengine/client/w3dview.cpp [hunk 6]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+    }
+
+    m_rcInfo.num_frames = milliseconds / g_theW3DFrameLengthInMsec;
+
+    if (m_rcInfo.num_frames < 1) {
+        m_rcInfo.num_frames = 1;
+    }
+
+    Coord3D pos = Get_Position();
+    Vector2 v1(p_loc->x - pos.x, p_loc->y - pos.y);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: transform

File: src/platform/w3dengine/client/w3dview.cpp [hunk 7]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
 
     Setup_Waypoint_Path(orient);
 }
+
+void W3DView::Rotate_Camera_Toward_Object(ObjectID id, int milliseconds, int hold_milliseconds, float in, float out)
+{
+    m_rcInfo.track_object = true;
+
+    if (hold_milliseconds < 1) {
+        hold_milliseconds = 0;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:1814]: 1, 0, correct?

File: src/platform/w3dengine/client/w3dview.cpp [hunk 8]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+    }
+
+    m_rcInfo.num_hold_frames = hold_milliseconds / g_theW3DFrameLengthInMsec;
+
+    if (m_rcInfo.num_hold_frames < 1) {
+        m_rcInfo.num_hold_frames = 0;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:1820]: 1, 0, correct?

File: src/platform/w3dengine/client/w3dview.cpp [hunk 9]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+    }
+
+    m_rcInfo.num_frames = milliseconds / g_theW3DFrameLengthInMsec;
+
+    if (m_rcInfo.num_frames < 1) {
+        m_rcInfo.num_frames = 1;
+    }
+
+    m_rcInfo.cur_frame = 0;
+    m_doingRotateCamera = 1;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: true

File: src/platform/w3dengine/client/w3dview.cpp [hunk 10]
```diff
@@ -38,6 +38,57 @@
...
+    m_locationRequests.reserve(50);
+    m_previousLookAtPosition.x = 0.0f;
+    m_previousLookAtPosition.y = 0.0f;
+    m_previousLookAtPosition.z = 0.0f;
+    m_scrollAmount.x = 0.0f;
+    m_scrollAmount.y = 0.0f;
+    m_cameraConstraint.hi.x = 0.0f;
+    m_cameraConstraint.hi.y = 0.0f;
+    m_cameraConstraint.lo.x = 0.0f;
+    m_cameraConstraint.lo.y = 0.0f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:83]: What about `m_rcInfo`, `m_pcInfo`, `m_zcInfo`, `m_mcwpInfo`? They are not fully initialized.

File: src/platform/w3dengine/client/w3dview.cpp [hunk 11]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+            Coord3D c2 = m_mcwpInfo.waypoints[i];
+            c2.x += m_mcwpInfo.waypoints[i + 1].x;
+            c2.y += m_mcwpInfo.waypoints[i + 1].y;
+            c2.x /= 2.0f;
+            c2.y /= 2.0f;
+            float x2 = (c2.x - x1) * 0.5f + x1;
+            float y2 = (c2.y - y1) * 0.5f + y1;
+            x2 += (1.0f - 0.5f) * 0.5f * (c1.x - c2.x + c1.x - x1);
+            y2 += (1.0f - 0.5f) * 0.5f * (c1.y - c2.y + c1.y - y1);
+            Vector2 v1(p_loc->x - x2, p_loc->y - y2);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: transform

File: src/platform/w3dengine/client/w3dview.cpp [hunk 12]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+            c3.x += m_mcwpInfo.waypoints[i + 1].x;
+            c3.y += m_mcwpInfo.waypoints[i + 1].y;
+            c3.x /= 2.0f;
+            c3.y /= 2.0f;
+
+            float x = (c3.x - c1.x) * 0.5f + c1.x;
+            float y = (c3.y - c1.y) * 0.5f + c1.y;
+            x += (1.0f - 0.5f) * 0.5f * (c2.x - c3.x + c2.x - c1.x);
+            y += (1.0f - 0.5f) * 0.5f * (c2.y - c3.y + c2.y - c1.y);
+            Vector2 v1(p_loc->x - x, p_loc->y - y);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: transform

File: src/platform/w3dengine/client/w3dview.cpp [hunk 13]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+        m_zcInfo.end_time_multiplier = final_multiplier;
+    }
+
+    if (m_doingPitchCamera) {
+        m_pcInfo.end_time_multiplier = final_multiplier;
+    }
+
+    if (m_doingRotateCamera) {
+        m_rcInfo.end_time_multiplier = final_multiplier;
+    } else if (m_doingMoveCameraOnWaypointPath) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:2084]: else if is correct?

File: src/platform/w3dengine/client/w3dview.cpp [hunk 14]
```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
...
+            Coord3D c1 = m_mcwpInfo.waypoints[i - 1];
+            c1.x += m_mcwpInfo.waypoints[i].x;
+            c1.y += m_mcwpInfo.waypoints[i].y;
+            c1.x /= 2.0f;
+            c1.y /= 2.0f;
+
+            Coord3D c2 = m_mcwpInfo.waypoints[i];
+            Coord3D c3 = m_mcwpInfo.waypoints[i];
+            c3.x += m_mcwpInfo.waypoints[i + 1].x;
+            c3.y += m_mcwpInfo.waypoints[i + 1].y;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:2038]: I don't quite understand how this is legal. Nothing appears to write to `m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints + 1]`, so how is this a valid read?

File: src/platform/w3dengine/client/w3dview.cpp (hunk 1 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:2184]: This is correct.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 3 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:2138]: Yes, blame EA for this.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 7 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:1814]: This is correct.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 8 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:1820]: This is correct.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 9 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 11 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 12 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 13 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:2084]: Yes.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 14 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp:2038]: This is correct, that variable is written to by Setup_Waypoint_Path.

**Review by xezon — APPROVED** [review]
Camera appears to behave as usual.

Shellmap looks as usual.
Mission camera intros work.
Camera zoom and blur work in Generals Challenge as usual.
Camera shake work as usual.
Camera controls work as usual.

No crashes.

**Pull request merged**